[13.0][ADD] base_ebill_payment_contract

TDu opened the pull request:
The goal of this module is to be a common ground for ebilling addons.
Alone this module has no purpose. It needs to be installed with a module
that implements a specific ebilling system.
This depending module has to implement his own 'transmit.method'.

TDu committed:
[13.0][ADD] base_ebill_payment_contract

The goal of this module is to be a common ground for ebilling addons.
Alone this module has no purpose. It needs to be installed with a module
that implements a specific ebilling system.
This depending module has to implement his own 'transmit.method'.

sebalix commented:
Depends on https://github.com/OCA/account-invoicing/pull/682

sebalix commented on base_ebill_payment_contract/models/ebill_payment_contract.py:
```diff
@@ -0,0 +1,53 @@
+
+    _name = "ebill.payment.contract"
+    _description = "Ebill Payment Contract"
+    _rec_name = "name"
```
```suggestion
```
As it's already the default

sebalix commented on base_ebill_payment_contract/models/ebill_payment_contract.py:
```diff
@@ -0,0 +1,53 @@
+    transmit_method_id = fields.Many2one(
+        comodel_name="transmit.method", string="Service Name", ondelete="restrict",
+    )
+    partner_id = fields.Many2one(
+        comodel_name="res.partner", required=True, string="Customer"
+    )
+    partner_id = fields.Many2one(
+        comodel_name="res.partner", required=True, string="Customer"
+    )
```
Field `partner_id` declared twice

sebalix commented on base_ebill_payment_contract/models/ebill_payment_contract.py:
```diff
@@ -0,0 +1,53 @@
+        It is valid if the contract is opened and his start date is in the pass
+        And his end date is in the future or not set.
+        """
+        for contract in self:
+            contract.is_valid = (
+                contract.state == "open"
+                and contract.date_start
+                and contract.date_start <= fields.Date.today()
+                and (not contract.date_end or contract.date_end >= fields.Date.today())
+            )
```
```suggestion
        today = fields.Date.today()
        for contract in self:
            contract.is_valid = (
                contract.state == "open"
                and contract.date_start
                and contract.date_start <= today
                and (not contract.date_end or contract.date_end >= today)
            )
```

sebalix commented on base_ebill_payment_contract/views/ebill_payment_contract.xml:
```diff
@@ -0,0 +1,64 @@
+                <field name="transmit_method_id" />
+                <field name="partner_id" />
+                <filter
+                    name="active"
+                    string="Active"
```
As `Active` has a different sense in Odoo, better to rename this `is_valid / Valid`?

TDu replied:
Has been changed.

sebalix reviewed (commented):
Small comments

sebalix commented:
You also have to add an `oca_dependencies.txt` to add `OCA/account-invoicing` repository with https://github.com/OCA/account-invoicing/pull/682

sebalix commented on base_ebill_payment_contract/views/ebill_payment_contract.xml:
```diff
@@ -0,0 +1,64 @@
+        <field name="arch" type="xml">
+            <form>
+                <sheet>
+                    <group>
```
```suggestion
                    <group name="base">
```
(or another name, I'm not inspired :p)

guewen commented on base_ebill_payment_contract/models/res_partner.py:
```diff
@@ -0,0 +1,28 @@
+        self.ensure_one()
+        # FIXME: Search on non stored field
+        contracts = self.env["ebill.payment.contract"].search(
+            [("is_valid", "=", True), ("partner_id", "=", self.id)], limit=1,
```
We have `Non-stored field ebill.payment.contract.is_valid cannot be searched.` in logs.

2 options:

* Implement a "search" method on the computed field
* Search all the contracts for the partner (remove limit) and filter them afterwards, it seems acceptable as the number of contract per partner will be a few. 

Solution 2 may be easier, but for UX, being able to search on the `is_valid` field may be useful if we have a list view with all the contracts.

TDu replied:
I implemented the search method on the third commit.

yvaucher commented on base_ebill_payment_contract/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Base Ebill Payment Contract",
```
```suggestion
    "name": "Base eBill Payment Contract",
```

yvaucher commented on base_ebill_payment_contract/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "Base Ebill Payment Contract",
+    "summary": """
+        Base for ebilling addons""",
```
```suggestion
        Base for e-billing addons""",
```

yvaucher commented on base_ebill_payment_contract/models/ebill_payment_contract.py:
```diff
@@ -0,0 +1,50 @@
+class EbillPaymentContract(models.Model):
+
+    _name = "ebill.payment.contract"
+    _description = "Ebill Payment Contract"
```
```suggestion
    _description = "eBill Payment Contract"
```

yvaucher commented on base_ebill_payment_contract/views/ebill_payment_contract.xml:
```diff
@@ -0,0 +1,64 @@
+        <field name="context">{}</field>
+    </record>
+    <record model="ir.ui.menu" id="ebill_payment_contract_menu">
+        <field name="name">Ebill Payment Contract</field>
```
```suggestion
        <field name="name">eBill Payment Contract</field>
```

TDu replied:
Thanks for the search on terminology, I fixed them all

yvaucher replied:
Seems this line slipped out of the fixing

eBill is still written with a lowercase at the start of a sentence.

An other orthography in "E-bill" with the dash. But most people have seen it as "eBill".

yvaucher reviewed (changes requested):
just a review about terminology.

Reference: https://www.ebill.ch/en/home.html

TDu committed:
[13.0][IMP] base_ebill_payment_contract search is_valid

Add the search implementation for the is_valid non stored computed
fields. So it works as expected.

yvaucher commented on base_ebill_payment_contract/views/ebill_payment_contract.xml:
```diff
@@ -0,0 +1,64 @@
+        </field>
+    </record>
+    <record model="ir.actions.act_window" id="ebill_payment_contract_act_window">
+        <field name="name">Ebill Payment Contract</field>
```
```suggestion
        <field name="name">eBill Payment Contract</field>
```

yvaucher commented on base_ebill_payment_contract/models/res_partner.py:
```diff
@@ -0,0 +1,28 @@
+    def get_active_contract(self, transmit_method):
+        """Return the active contract for a specific transmit method."""
+        self.ensure_one()
+        # FIXME: Search on non stored field
```
You can remove this FIXME if it was for `is_valid`

TDu replied:
Done

yvaucher commented on base_ebill_payment_contract/models/res_partner.py:
```diff
@@ -0,0 +1,28 @@
+                    self.name, transmit_method.name
+                )
+            )
+        return contracts[0] if contracts else contracts
```
if you name your variable `contract` without the `s` doesn't it make it more obvious you are doing a search with `limit=1` ?

yvaucher reviewed (approved):
Other than the 2 remaining "Ebill" terms to change to "eBill" and a couple of nitpicking

 LGTM

jgrandguillaume reviewed (approved):
Functional test ok

yvaucher commented:
/ocabot merge nobump
